Create quest_user_status view showing students’ submission statuses; Closes #918

thefoxwithin opened the pull request:
Please ensure you are familiar with our Pull Request Description expectations described here: https://www.pullrequest.com/blog/writing-a-great-pull-request-description/
### What?
Add a new "Quest User Status" view, URL route, and template to display the status of all active students for a given quest. The page shows each student's latest submission status—such as Not Started, Completed, Returned, Awaiting Approval, or In Progress—with links to submission details when available. Also add a button linking to this status page where appropriate.

### Why?
This feature provides teachers and staff a centralized view of student progress on quests, improving oversight and communication regarding quest submissions and statuses.

https://github.com/bytedeck/bytedeck/issues/918

### How?
- Added `quest_user_status` view decorated with `@staff_member_required` that fetches the quest, active student profiles, their latest submissions, and computes each user’s submission status.
- Created a URL route at `r'^(?P<quest_id>[0-9]+)/user-status/$'` named `quest_user_status` for accessing this view.
- Created the template `quest_user_status.html` extending `quest_manager/base.html`, displaying quest info and a sortable/searchable table of users with their statuses.
- Added a conditional button link to `quest_user_status` in quest-related templates, excluding the page itself to avoid redundant navigation.

### Testing

Added unit tests for the `quest_user_status` view.

Verified correct display of user submission statuses, including:

- Users with approved submissions show as **Approved**.  
- Users with returned submissions show as **Returned**.  
- Users with pending submissions show as **Awaiting Approval**.  
- Users with no submissions show as **Not Started**.

Ensured the view responds with HTTP 200 status for all cases tested.

Test coverage includes multiple users with different submission states.

### Screenshots (if front end is affected)

https://github.com/user-attachments/assets/c962ddcf-7a98-4afb-8bda-ff14bd2a17fc


### Anything Else?
Planning on adding the semester tracking like the Badge view, current semester and all semesters options. That might have to come later though.

### Review request
@tylerecouture


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Staff can open a "View status of all students" action from quest pages to see a completion dashboard with quest details (XP, campaign, expiry, repeatability), related maps and tags, prerequisites, max total XP, a status breakdown (counts & percentages), and a per-student table with links to submissions.

* **Tests**
  * Added tests for the status dashboard covering access, per-user statuses, "not started" cases, and latest-attempt selection.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tylerecouture reviewed (approved):
Amazing!

tylerecouture reviewed (changes requested):
Instead of this upper formatting, let's copy the summary/stats page (e.g. https://hackerspace.bytedeck.com/quests/322/summary/)

With "Completion Statuses" as the main titles and then Quest: [Quest Name with link] then hr.  Could we add some basic completion stats?  Like:
Approved: 5 (10%)
Awaiting Approval: 2 (4%)
In Progress:  20 (40%)
Not Started: 23 (46%)

Also, instead of using "Completed" which can be confusing and was a terrible name choice on my part, use "Awaiting Approval" on the front end.

<img width="881" height="418" alt="image" src="https://github.com/user-attachments/assets/eb7c503f-66eb-4208-8d1e-958df143f8cb" />

thefoxwithin committed:
Create quest_user_status view showing students’ submission statuses

Add user status button to quest action buttons

Add template to display quest info and user status table with links

Add URL route for quest_user_status view

minimal tests

Add unit tests for QuestUserStatus view covering user statuses

fixed bugs caught by coderabbit

added multiple submission ordering test

added statistics for submissions

removed syntax errors added more explicit testing

change status name from completed to Approved

added ordering and fixed tests

tylerecouture reviewed (changes requested):
"Users with approved submissions show as Completed." <-- Approved submissions should show as "Approved"

Also, please update the screenshots/video, I think it's out of date now?  It's confusing me.

thefoxwithin commented:
> "Users with approved submissions show as Completed." <-- Approved submissions should show as "Approved"
> 
> Also, please update the screenshots/video, I think it's out of date now? It's confusing me.

Yeah I made the changes but forgot to update the examples I gave. My bad oops.

tylerecouture reviewed (changes requested):
The order here seems random, 

<img width="547" height="315" alt="image" src="https://github.com/user-attachments/assets/03b8f29c-fed0-4200-80a2-d18008723e52" />

can you order them logically like this:

Approved:
Returned:
Awaiting Approval:
In Progress:
Not Started:

tylerecouture reviewed (changes requested):
In the video, after you clicked to sort via completion status, it didn't sort properly:

<img width="540" height="416" alt="image" src="https://github.com/user-attachments/assets/50640794-3b8e-45b3-ac21-ec115ff9970d" />

thefoxwithin committed:
Merge branch 'develop' into quest-status-view

tylerecouture commented:
> Now we're sorting the tables in the view, and since it's being sorted beforehand the status part of the table is no longer "sortable" since that would just break the sorting.

Why doesn't the table's sorting work?  isn't it just alpha numeric?  or was it set to sort on the wrong column?

tylerecouture commented:
> > Now we're sorting the tables in the view, and since it's being sorted beforehand the status part of the table is no longer "sortable" since that would just break the sorting.
> 
> Why doesn't the table's sorting work? isn't it just alpha numeric? or was it set to sort on the wrong column?

Oh, I know why, it's because of the URL, it's sorting on the HTML not the displayed text.  That's an easy fix we have to do in various places in the app.  Easiest way is put the raw text in an invisible column and sort on the raw column instead

tylerecouture commented:
> > > Now we're sorting the tables in the view, and since it's being sorted beforehand the status part of the table is no longer "sortable" since that would just break the sorting.
> > 
> > 
> > Why doesn't the table's sorting work? isn't it just alpha numeric? or was it set to sort on the wrong column?
> 
> Oh, I know why, it's because of the URL, it's sorting on the HTML not the displayed text. That's an easy fix we have to do in various places in the app. Easiest way is put the raw text in an invisible column and sort on the raw column instead

Or see this: https://examples.bootstrap-table.com/#options/custom-sort.html#view-source
